File: frontend/src/utils/trackerSettings.js
import { normalizeExternalHttpUrl } from './textSanitization'

export const TRACKER_TOOL_ACCESS_OPTIONS = [
  { value: 'admin', label: 'Admin' },
  { value: 'team', label: 'Team' },
  { value: 'all', label: 'All' },
]

export const DEFAULT_DELIVERY_MESSAGE = 'Thanks for reviewing with Vue.'

const TRACKER_TOOL_ACCESSES = new Set(TRACKER_TOOL_ACCESS_OPTIONS.map(option => option.value))

function cleanDeliveryLogoUploadName(value) {
  const name = String(value || '').trim().split(/[\\/]/).pop() || ''
  if (!name.startsWith('delivery-logo-')) return ''
  if (!/\.(jpe?g|png|gif|webp|bmp|tiff?|hei[cf])$/i.test(name)) return ''
  return name
}

export function normalizeDeliveryLinks(value) {
  if (!Array.isArray(value)) return []
  const links = []
  for (const item of value) {
    if (!item || typeof item !== 'object') continue
    const label = String(item.label || '').trim().slice(0, 48)
    const url = normalizeExternalHttpUrl(item.url)
    if (!label || !url) continue
    links.push({ label, url })
    if (links.length >= 4) break
  }
  return links
}

export function normalizeTrackerSettings(rawSettings = null, options = {}) {
  const preserveDeliveryMessage = options?.preserveDeliveryMessage === true
  const tools = {
    comparison: {
      enabled: true,
      access: 'team',
    },
    details: {
      enabled: true,
      access: 'all',
    },
    brief_preview: {
      enabled: true,
    },
    version_review: {
      enabled: false,
    },
    delivery: {
      enabled: false,
      message: DEFAULT_DELIVERY_MESSAGE,
      notes: '',
      links: [],
      logo_upload_name: '',
    },
  }

  for (const key of ['comparison', 'details']) {
    const item = rawSettings?.[key]
    if (!item || typeof item !== 'object') continue
    if ('enabled' in item) tools[key].enabled = item.enabled !== false
    if ('access' in item && TRACKER_TOOL_ACCESSES.has(String(item.access || '').toLowerCase())) {
      tools[key].access = String(item.access).toLowerCase()
    } else if (key === 'comparison' && 'share_access' in item) {
      tools[key].access = item.share_access === true ? 'all' : 'team'
    }
  }

  const briefPreview = rawSettings?.brief_preview
  if (briefPreview && typeof briefPreview === 'object' && 'enabled' in briefPreview) {
    tools.brief_preview.enabled = briefPreview.enabled !== false
  }

  const versionReview = rawSettings?.version_review
  if (versionReview && typeof versionReview === 'object' && 'enabled' in versionReview) {
    tools.version_review.enabled = versionReview.enabled === true
  }

  const delivery = rawSettings?.delivery
  if (delivery && typeof delivery === 'object') {
    if ('enabled' in delivery) tools.delivery.enabled = delivery.enabled === true
    if ('message' in delivery) {
      if (preserveDeliveryMessage) {
        tools.delivery.message = String(delivery.message ?? '')
      } else {
        const message = String(delivery.message || '').trim()
        tools.delivery.message = message || DEFAULT_DELIVERY_MESSAGE
      }
    }
    if ('notes' in delivery) {
      tools.delivery.notes = preserveDeliveryMessage
        ? String(delivery.notes ?? '')
        : String(delivery.notes || '').trim().slice(0, 1200)
    }
    if ('links' in delivery) {
      tools.delivery.links = preserveDeliveryMessage && Array.isArray(delivery.links)
        ? delivery.links.slice(0, 4).map(link => ({
            label: String(link?.label ?? ''),
            url: String(link?.url ?? ''),
          }))
        : normalizeDeliveryLinks(delivery.links)
    }
    if ('logo_upload_name' in delivery) {
      tools.delivery.logo_upload_name = cleanDeliveryLogoUploadName(delivery.logo_upload_name)
    }
  }

  return tools
}

export function trackerToolEnabledForContext(tracker, toolKey, { shareMode = false, currentUser = null, accessRole = null } = {}) {
  const tool = normalizeTrackerSettings(tracker?.settings)[toolKey]
  if (!tool?.enabled) return false
  if (toolKey === 'delivery') return true

  if (tool.access === 'all') return true
  if (shareMode) return false
  if (tool.access === 'admin') return currentUser?.role === 'admin'
  return Boolean(accessRole || currentUser)
}

File: backend/app/services/horizons/tracker_settings.py
from __future__ import annotations

import json

from app.models import HorizonTracker
from app.services.external_urls import normalize_external_http_url
from app.services.project_delivery import normalize_delivery_logo_upload_name

from .common import DEFAULT_DELIVERY_MESSAGE, DEFAULT_TRACKER_SETTINGS, TRACKER_TOOL_ACCESSES

def _normalize_delivery_links(value) -> list[dict[str, str]]:
    if not isinstance(value, list):
        return []
    links = []
    for item in value:
        if not isinstance(item, dict):
            continue
        label = str(item.get('label') or '').strip()[:48]
        url = normalize_external_http_url(item.get('url'))
        if not label or not url:
            continue
        links.append({'label': label, 'url': url})
        if len(links) >= 4:
            break
    return links


def normalize_tracker_settings(value: dict | None) -> dict:
    tools = json.loads(json.dumps(DEFAULT_TRACKER_SETTINGS))
    if not isinstance(value, dict):
        return tools

    for key in ('comparison', 'details'):
        item = value.get(key)
        if not isinstance(item, dict):
            continue
        if 'enabled' in item:
            tools[key]['enabled'] = bool(item.get('enabled'))
        if 'access' in item:
            access = str(item.get('access') or '').strip().lower()
            if access in TRACKER_TOOL_ACCESSES:
                tools[key]['access'] = access
        elif key == 'comparison' and 'share_access' in item:
            tools[key]['access'] = 'all' if item.get('share_access') is True else 'team'

    brief_preview = value.get('brief_preview')
    if isinstance(brief_preview, dict) and 'enabled' in brief_preview:
        tools['brief_preview']['enabled'] = bool(brief_preview.get('enabled'))

    version_review = value.get('version_review')
    if isinstance(version_review, dict) and 'enabled' in version_review:
        tools['version_review']['enabled'] = bool(version_review.get('enabled'))

    delivery = value.get('delivery')
    if isinstance(delivery, dict):
        if 'enabled' in delivery:
            tools['delivery']['enabled'] = bool(delivery.get('enabled'))
        if 'message' in delivery:
            message = str(delivery.get('message') or '').strip()
            tools['delivery']['message'] = message or DEFAULT_DELIVERY_MESSAGE
        if 'notes' in delivery:
            tools['delivery']['notes'] = str(delivery.get('notes') or '').strip()[:1200]
        if 'links' in delivery:
            tools['delivery']['links'] = _normalize_delivery_links(delivery.get('links'))
        if 'logo_upload_name' in delivery:
            tools['delivery']['logo_upload_name'] = normalize_delivery_logo_upload_name(delivery.get('logo_upload_name'))
    return tools


def tracker_settings_for(tracker: HorizonTracker) -> dict:
    try:
        raw_tools = json.loads(tracker.settings_json or '{}')
    except Exception:
        raw_tools = {}
    return normalize_tracker_settings(raw_tools)


def tracker_tool_enabled_for_context(
    tracker: HorizonTracker,
    tool_key: str,
    *,
    user: dict | None = None,
    access_role: str | None = None,
    share: bool = False,
) -> bool:
    tool = tracker_settings_for(tracker).get(tool_key) or {}
    if not tool.get('enabled'):
        return False
    if tool_key == 'delivery':
        return True

    access = str(tool.get('access') or 'team').strip().lower()
    if access == 'all':
        return True
    if share:
        return False
    if access == 'admin':
        return bool(user and user.get('role') == 'admin')
    return bool(access_role)
